### PR TITLE
KAFKA-13495: Add reason to JoinGroupRequest (KIP-800)

### PR DESCRIPTION
**Pull request by jeffkbkim**
As part of [KIP-800](https://cwiki.apache.org/confluence/display/KAFKA/KIP-800%3A+Add+reason+to+JoinGroupRequest+and+LeaveGroupRequest), this patch aims to add a reason to why a consumer is joining a group and log it in the broker. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/resources/common/message/JoinGroupRequest.json [hunk 1]
```diff
@@ -30,7 +30,9 @@
   // Version 6 is the first flexible version.
   //
   // Version 7 is the same as version 6.
```
Comment (socutes) [inline clients/src/main/resources/common/message/JoinGroupRequest.json:32]: It would be better to keep the version's change notes here rather than delete them

Comment (jeffkbkim) [reply clients/src/main/resources/common/message/JoinGroupRequest.json:32]: thanks for the comment. 
these lines aren't being removed. i think github ui is marking red for json comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -2303,18 +2303,24 @@ public ConsumerGroupMetadata groupMetadata() {
      * @throws java.lang.IllegalStateException if the consumer does not use group subscription
      */
     @Override
-    public void enforceRebalance() {
+    public void enforceRebalance(final String reason) {
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:2317]: Why does the new API not have new java doc for it?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 3]
```diff
@@ -181,6 +182,7 @@ class GroupCoordinator(val brokerId: Int,
           responseCallback(JoinGroupResult(memberId, Errors.UNKNOWN_MEMBER_ID))
         case Some(group) =>
           group.inLock {
+            info(s"memberId=$memberId with groupInstanceId=$groupInstanceId is attempting to join groupId=$groupId due to: $reason")
```
Comment (showuon) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Since the consumer might be in the old version, the `reason` might be `null`. I think we should not output the log when the `reason` is null. i.e.:
`memberId=1 with groupInstanceId=2 is attempting to join groupId=3 due to: null`

**Review by showuon — COMMENTED** [review]
@jeffkbkim , thanks for the PR. Overall looks good. Left some comments. Thanks.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 4]
```diff
@@ -2300,21 +2300,32 @@ public ConsumerGroupMetadata groupMetadata() {
...
         acquireAndEnsureOpen();
         try {
             if (coordinator == null) {
                 throw new IllegalStateException("Tried to force a rebalance but consumer does not have a group.");
             }
-            coordinator.requestRejoin("rebalance enforced by user");
+            String defaultReason = "rebalance enforced by user";
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: nit: Should we define this one as a constant?

File: clients/src/main/resources/common/message/JoinGroupRequest.json [hunk 5]
```diff
@@ -30,7 +30,9 @@
...
-  "validVersions": "0-7",
+  //
+  // Version 8 adds the Reason field (KIP-800).
+  "validVersions": "0-8",
```
Comment (dajac) [inline clients/src/main/resources/common/message/JoinGroupRequest.json:35]: We need to bump the request version as well.

File: clients/src/main/resources/common/message/JoinGroupRequest.json [hunk 6]
```diff
@@ -54,6 +56,9 @@
         "about": "The protocol name." },
       { "name": "Metadata", "type": "bytes", "versions": "0+",
         "about": "The protocol metadata." }
-    ]}
+    ]},
+    { "name": "Reason", "type": "string",
+      "versions": "8+", "nullableVersions": "8+","default": "null",
```
Comment (dajac) [inline clients/src/main/resources/common/message/JoinGroupRequest.json]: nit: Could we put `versions` on the previous line to stay inline with the other fields? A space is missing before `default`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java [hunk 7]
```diff
@@ -541,8 +579,17 @@ private void expectDisconnectInJoinGroup(
         }, null, true);
     }
 
+    private void expectJoinGroup(
+            String expectedMemberId,
+            int responseGeneration,
+            String responseMemberId
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: nit: Indentation is off.

File: clients/src/main/resources/common/message/JoinGroupRequest.json (hunk 6 above)
Comment (dajac) [reply clients/src/main/resources/common/message/JoinGroupRequest.json]: Could we also update `RequestResponseTest` to cover this change?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 8]
```diff
@@ -133,6 +133,7 @@ public boolean hasNotJoinedGroup() {
     protected final ConsumerNetworkClient client;
 
     private Node coordinator = null;
+    private String rejoinReason = "initialized abstract coordinator";
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: The reason is a bit weird. I wonder if we could just leave it empty in the beginning.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 9]
```diff
@@ -467,6 +468,7 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 final RuntimeException exception = future.exception();
 
                 resetJoinGroupFuture();
+                rejoinReason = exception.getMessage();
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I wonder if we should put a bit more information here. For instance, we could say `rebalance failed due to %s error: %s` where the first `%s` would be the exception's class and the second the message. What do you think?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 10]
```diff
@@ -163,6 +163,7 @@ class GroupCoordinator(val brokerId: Int,
                       protocolType: String,
                       protocols: List[(String, Array[Byte])],
                       responseCallback: JoinCallback,
+                      reason: String = null,
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: I would rather use `Option[String]` here.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 11]
```diff
@@ -181,6 +182,8 @@ class GroupCoordinator(val brokerId: Int,
           responseCallback(JoinGroupResult(memberId, Errors.UNKNOWN_MEMBER_ID))
         case Some(group) =>
           group.inLock {
+            if (reason != null)
+              info(s"memberId=$memberId with groupInstanceId=$groupInstanceId is attempting to join groupId=$groupId due to: $reason")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: We already log a few messages when a member (re-)joins. I wonder if we should pass the reason down and use it in all the existing messages that we already have. What do you think? I need to look at this a bit more myself.

**Review by dajac — COMMENTED** [review]
@jeffkbkim Thanks for the PR. I left a few comments.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java [hunk 12]
```diff
@@ -486,6 +487,43 @@ public void testRetainMemberIdAfterSyncGroupDisconnect() {
...
+        coordinator.requestRejoin("Manual test trigger");
+        ensureActiveGroup(generation, memberId);
+        assertEquals("", coordinator.rejoinReason());
+
+        // max group size reached
+        mockClient.prepareResponse(joinGroupFollowerResponse(defaultGeneration, memberId, JoinGroupRequest.UNKNOWN_MEMBER_ID, Errors.GROUP_MAX_SIZE_REACHED));
+        coordinator.requestRejoin("Manual test trigger 2");
+        try {
+            coordinator.joinGroupIfNeeded(mockTime.timer(100L));
+        } catch (GroupMaxSizeReachedException e) {
```
Comment (showuon) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: any reason why we don't use `assertThrows` here?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java [hunk 13]
```diff
@@ -559,8 +606,10 @@ private void expectJoinGroup(
                 return false;
             }
             JoinGroupRequestData joinGroupRequest = ((JoinGroupRequest) body).data();
+            boolean isReasonMatching = expectedReason == null || joinGroupRequest.reason().equals(expectedReason);
```
Comment (showuon) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: I don't think the logic here is correct. 
If the `expectedReason` is null, but the `joinGroupRequest.reason()` is not null, should it be passed?
That also implies you need add tests to cover this situation.

**Review by showuon — COMMENTED** [review]
@jeffkbkim , thanks for the update. Left some comments.

File: clients/src/main/resources/common/message/JoinGroupRequest.json (hunk 5 above)
Comment (jeffkbkim) [reply clients/src/main/resources/common/message/JoinGroupRequest.json:35]: bumped the response version

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 11 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: yeah, that was my original thoughts as well. i believe you are referring to `maybePrepareRebalance(group: GroupMetadata, reason: String)`. 
methods that are referencing this are:
- `addMemberAndRebalance`
- `doSyncGroup` (NA)
- `removeMemberAndUpdateGroup` (NA)
- `updateMemberAndRebalance`
- `updateStaticMemberAndRebalance`

i've passed the reason into all parent methods. also, i've updated `updateStaticMemberAndRebalance`'s direct call to `prepareRebalance()`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java (hunk 12 above)
Comment (jeffkbkim) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: we do not want to stop when an exception is thrown - we need to verify that the reason is updated to the appropriate exception message.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java (hunk 13 above)
Comment (jeffkbkim) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: i added a comment. the reason is never set to null in AbstractCoordinator.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 14]
```diff
@@ -181,6 +182,7 @@ class GroupCoordinator(val brokerId: Int,
           responseCallback(JoinGroupResult(memberId, Errors.UNKNOWN_MEMBER_ID))
         case Some(group) =>
           group.inLock {
+            val joinReason = reason.getOrElse("unknown reason")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: If we do this, it might be better to not use an `Option` after all. We could simply provided the default reason to `handleJoinGroup` if none is provided. Also, how about using `not provided` instead of `unknown reason`?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 15]
```diff
@@ -1249,7 +1260,7 @@ class GroupCoordinator(val brokerId: Int,
     // for new members. If the new member is still there, we expect it to retry.
     completeAndScheduleNextExpiration(group, member, NewMemberJoinTimeoutMs)
 
-    maybePrepareRebalance(group, s"Adding new member $memberId with group instance id $groupInstanceId")
+    maybePrepareRebalance(group, s"Adding new member $memberId with group instance id $groupInstanceId. Member joined due to $reason")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: The output log is quite hard to follow at the moment. Example:

```
[2021-12-17 11:29:16,061] INFO [GroupCoordinator 0]: Preparing to rebalance group test in state PreparingRebalance with old generation 1 (__consumer_offsets-48) (reason: Adding new member console-consumer-1d5a9905-c271-4700-a817-62fc9b9f28fc with group instance id None. Member joined due to rebalance failed due to class org.apache.kafka.common.errors.MemberIdRequiredException error: The group member needs to have a valid member id before actually entering a consumer group.) (kafka.coordinator.group.GroupCoordinator)
```

How about doing the following? For each reason, we could add `; client reason: $reason`. With this, we will always have (reason: ....; client reason: ...) in each rebalance logs. It might be clearer. What do you think?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 16]
```diff
@@ -467,6 +468,7 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 final RuntimeException exception = future.exception();
 
                 resetJoinGroupFuture();
+                rejoinReason = "rebalance failed due to " + exception.getClass() + " error: " + exception.getMessage();
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Example on the broker side:
```
[2021-12-17 11:29:16,061] INFO [GroupCoordinator 0]: Preparing to rebalance group test in state PreparingRebalance with old generation 1 (__consumer_offsets-48) (reason: Adding new member console-consumer-1d5a9905-c271-4700-a817-62fc9b9f28fc with group instance id None. Member joined due to rebalance failed due to class org.apache.kafka.common.errors.MemberIdRequiredException error: The group member needs to have a valid member id before actually entering a consumer group.) (kafka.coordinator.group.GroupCoordinator)
```

* Should we only get the `getSimpleName` of the class?
* There are many `:` in the log. I wonder if we could remove the one we've put here. Perhaps, we could use the following pattern: `rebalance failed due to '$message' ($class)`. What do you think?

File: clients/src/main/resources/common/message/JoinGroupResponse.json [hunk 17]
```diff
@@ -31,7 +31,9 @@
   // Version 6 is the first flexible version.
   //
   // Starting from version 7, the broker sends back the Protocol Type to the client (KIP-559).
-  "validVersions": "0-7",
+  //
+  // Version 8 adds the Reason field (KIP-800).
```
Comment (dajac) [inline clients/src/main/resources/common/message/JoinGroupResponse.json]: nit: Should we rather say Version 8 is the same as version 7. here?

**Review by dajac — COMMENTED** [review]
@jeffkbkim Thanks for the update. I left a few more comments.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java (hunk 12 above)
Comment (showuon) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: Actually, you can achieve what you want by `assertThrows` as below:

```java
        Throwable e = assertThrows(GroupMaxSizeReachedException.class,
            () -> coordinator.joinGroupIfNeeded(mockTime.timer(100L)));

        // next join group request should contain exception message
        expectJoinGroup(memberId, e.getMessage(), generation, memberId);
        expectSyncGroup(generation, memberId);
        ensureActiveGroup(generation, memberId);
        assertEquals("", coordinator.rejoinReason());
```

Basically, the `assertThrows` is doing the similar thing as what you did here (try/catch). It's recommended to use the `assertThrows` to understand the exception is also a verification position in this test. Otherwise, let's say, if someday, someone breaks the logic and makes the `coordinator.joinGroupIfNeeded(mockTime.timer(100L));` works without exception thrown, your current test can't catch this error, right?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 15 above)
Comment (showuon) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: +1 for `; client reason: $reason`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 16 above)
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: +1 for `getSimpleName` for the class. In addition to the David's suggestion, I think we should also remove the 2nd `due to`, because there is already 1 `due to` in the sentence. ex:  `rebalance failed: '$message' ($class)`

**Review by showuon — COMMENTED** [review]
@jeffkbkim , thanks for your update. Left some comments. Thanks.

**Review by showuon — COMMENTED** [review]
Also, there are some failed tests. Please take a look. Thanks.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java (hunk 12 above)
Comment (jeffkbkim) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: ah gotcha. makes sense, thanks!

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 14 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: good idea! i'll update the leave group as well.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 16 above)
Comment (jeffkbkim) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: thanks for the suggestions. i'll go ahead with david's suggestion as we're updating `Member joined due to` to `client reason:`

**Review by showuon — COMMENTED** [review]
@jeffkbkim , thanks for the update. LGTM. Please fix the compilation error. Thanks.

**Comment by jeffkbkim** [comment]
@dajac @showuon i reverted to using `reason: Option[String]` for `handleJoinGroup()`. KafkaApi always pass in a `reason` which could be `null` so we would see "null" as the reason instead of "not provided".

**Comment by jeffkbkim** [comment]
@dajac gentle reminder to take another pass

**Comment by showuon** [comment]
@jeffkbkim , `KafkaApisTest` is still failed some tests. Please take a look. Thanks.

https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11566/6/#showFailuresLink

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 18]
```diff
@@ -467,6 +468,7 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 final RuntimeException exception = future.exception();
 
                 resetJoinGroupFuture();
+                rejoinReason = "rebalance failed due to '" + exception.getMessage() + "' (" + exception.getClass().getSimpleName() + ")";
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: nit: Would it make sense to use `String.format` like we did at L460?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 19]
```diff
@@ -1690,6 +1690,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         joinGroupRequest.data.protocolType,
         protocols,
         sendResponseCallback,
+        Option(joinGroupRequest.data.reason()),
```
Comment (dajac) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit: You can remove the parenthesis after `reason`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 18 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Also, it might be better to use `synchronized (AbstractCoordinator.this) { }` to mutate both `rejoinReason` and `rejoinNeeded` in order to ensure that they are consistent with each others.

**Review by dajac — COMMENTED** [review]
@jeffkbkim Thanks for the update. I left a few minor comments.

**Review by dajac — APPROVED** [review]
LGTM, thanks!
